vision/pipeline.py

```diff
@@ -3,62 +3,114 @@
+    flight_communication: multiprocessing queue
+        Interface to recieve flight state information from flight.
+    camera: Camera
+        Camera to pull image from.
+    """
+    PUT_TIMEOUT = 1  # Expected time for results to be irrelevant.
 
-    env: Environment
-        The environment interface that is used by flight code.
-        The pipeline updates the interface.
+    def __init__(self, vision_communication, flight_communication, camera):
+        ##
+        self.vision_communication = vision_communication
+        self.flight_communication = flight_communication
 
-    alg_time: int
-        An integer value that corresponds to how long the video loops.
-    """
-    def __init__(self, vid_file, env, config, alg_time=98):
-        self.vid_file = vid_file
-        self.env = env
-        self.alg_time = alg_time
-        self.config = config
+        self.camera = camera.__iter__()
+
+        ##
+        prefix = 'vision' if os.path.isdir("vision") else ''
 
-    def run_algorithm(self):
+        #
+        config_filename = os.path.join(prefix, 'obstacle', 'config.json')
+
+        with open(config_filename, 'r') as config_file:
+            config = json.load(config_file)
+
+        self.obstacle_finder = ObstacleFinder(params=import_params(config))
+
+    @property
+    def picture(self):
+        return next(self.camera)
+
+    def run(self, prev_state):
         """
-        Method that takes the given video file and environment, and updates the
-        environment with detected blobs.
+        Process current camera frame.
         """
-        for i, (depth_image, color_image) in enumerate(BagFile(0, 0, 0, self.vid_file)):
-            if i == self.alg_time:
-                break
-            obstacle_finder = ObstacleFinder(params=import_params.import_params(self.config))
-            bboxes = obstacle_finder.find(color_image, depth_image)
-            self.env.update(bboxes)
+        ##
+        depth_image, color_image = self.picture
+
+        try:
+            state = self.flight_communication.get_nowait()
+        except Empty:
+            state = prev_state
 
-            #from util.obstacle_plotter import plot_obstacles
-            #plot_obstacles(obstacle_finder.keypoints, color_image)
+        ##
+        bboxes = []
+
+        if state == 'early_laps':
+            bboxes = self.obstacle_finder.find(color_image, depth_image)
+        else:
+            pass  # raise AttributeError(f"Unrecognized state: {state}")
+
+        ##
+        for bbox in bboxes:
+            self.vision_communication.put(bbox, self.PUT_TIMEOUT)
+
+        # from vision.common.blob_plotter import plot_blobs
+        # plot_blobs(self.obstacle_finder.keypoints, color_image)
+
+        return state
+
+    def call_me(self):
+        """
+        Alex, call this function - not run.
+        """
+        prev_state = 'start'
+
+        for _ in range(100):
+            prev_state = pipeline.run(prev_state)
```

Comment:
Would prefer if this was not part of the class, and created an instance of the class and called its run function. You also have a magic number here, 100 calls? Why not just go forever

Reply:
Good idea, the reason it does not go forever is because the bag file reader will just endlessly iterate over videos. It is nice to have it terminate automatically.